### PR TITLE
Integrating AIR with the ROCm runtime

### PR DESCRIPTION
**Pull request by eddierichter-amd**

### CODE REVIEW

**Review by fifield — APPROVED** [review]

File: .github/workflows/buildAndTest.yml [hunk 1]
```diff
@@ -27,6 +27,27 @@ jobs:
           fetch-depth: 2
           submodules: "true"
 
+      - name: Install necessary build tools for libelf
```
Comment (jnider) [inline .github/workflows/buildAndTest.yml:30]: We recently decided to replace our custom libelf with the pre-built one. It means these tools are no longer necessary. I don't know if it's better to make that change in a different commit or to edit this one. What do you think?

File: .github/workflows/buildAndTest.yml [hunk 2]
```diff
@@ -66,12 +87,26 @@ jobs:
...
 
+      - name: Get ROCm-air-platforms
+        id: clone-rocm-air-platforms
+        run: utils/clone-rocm-air-platforms.sh
+        shell: bash
+
       - name: Rebuild and Install libxaie
         run: utils/github-clone-build-libxaie.sh
 
+      - name: Rebuild and Install elfutils
```
Comment (jnider) [inline .github/workflows/buildAndTest.yml:98]: Also part of the libelf change

File: .github/workflows/buildAndTest.yml [hunk 3]
```diff
@@ -86,7 +121,10 @@ jobs:
             -DMLIR_DIR=../llvm/install/lib/cmake/mlir/ \
             -DLLVM_DIR=../llvm/install/lib/cmake/llvm/ \
             -DAIE_DIR=`pwd`/../mlir-aie/install/lib/cmake/aie/ \
+            -DELFUTILS_DIR=$PWD/../elfutils \
```
Comment (jnider) [inline .github/workflows/buildAndTest.yml:124]: Also part of the libelf change

File: .github/workflows/buildAndTest.yml [hunk 4]
```diff
@@ -113,7 +151,10 @@ jobs:
...
```
Comment (jnider) [inline .github/workflows/buildAndTest.yml:154]: Also part of the libelf change

File: CMakeLists.txt [hunk 5]
```diff
@@ -150,7 +150,10 @@ foreach(target ${AIR_RUNTIME_TARGETS})
                -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
                -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}
                -DLibXAIE_ROOT=${LibXAIE_ROOT}
+               -Dhsa-runtime64_DIR=${hsa-runtime64_DIR}
+               -Dhsakmt_DIR=${hsakmt_DIR}
                -DAIE_DIR=${AIE_DIR}
+               -DELFUTILS_DIR=${ELFUTILS_DIR}
```
Comment (jnider) [inline CMakeLists.txt:156]: Also part of the libelf change

File: docs/building.md [hunk 6]
```diff
@@ -46,127 +64,51 @@ First, run utils/setup_python_packages.sh to setup the prerequisite python packa
...
-cd utils
-./clone-llvm.sh
-./build-llvm-local.sh llvm build ../../install
+cd mlir-air
+./utils/clone-llvm.sh
+./utils/build-llvm-local.sh llvm
 ```
 
-Next, clone and build MLIR-AIE with paths to llvm, aienginev2, and cmakeModules repositories. Again, we use a common installation directory. Note that in the following commands, we assume that the aienginev2 library is installed in /opt/xaiengine as directed in the `Building on x86 with runtime for PCIe` section. If the aienginev2 library was installed elsewhere, be sure that the 4th argument to build mlir-aie points to that location. 
+Next, we are going to need to clone and build elfutils. This is used in the runtime to unpack and decode AIE application binaries.
```
Comment (jnider) [inline docs/building.md:77]: Also part of the libelf change

File: docs/building.md [hunk 7]
```diff
@@ -46,127 +64,51 @@ First, run utils/setup_python_packages.sh to setup the prerequisite python packa
...
+./utils/clone-llvm.sh
+./utils/build-llvm-local.sh llvm
 ```
 
-Next, clone and build MLIR-AIE with paths to llvm, aienginev2, and cmakeModules repositories. Again, we use a common installation directory. Note that in the following commands, we assume that the aienginev2 library is installed in /opt/xaiengine as directed in the `Building on x86 with runtime for PCIe` section. If the aienginev2 library was installed elsewhere, be sure that the 4th argument to build mlir-aie points to that location. 
+Next, we are going to need to clone and build elfutils. This is used in the runtime to unpack and decode AIE application binaries. 
 ```
-./clone-mlir-aie.sh
-./build-mlir-aie-local.sh llvm mlir-aie/cmake/modulesXilinx /opt/xaiengine mlir-aie build ../../install
+./utils/github-clone-build-elfutils.sh
```
Comment (jnider) [inline docs/building.md:79]: Also part of the libelf change

File: runtime_lib/airhost/CMakeLists.txt [hunk 8]
```diff
@@ -30,16 +34,29 @@ set_property(TARGET airhost PROPERTY POSITION_INDEPENDENT_CODE ON)
...
     queue.cpp
+    runtime.cpp
     host.cpp
     pcie-ernic.cpp
     pcie-ernic-dev-mem-allocator.cpp
     network.cpp
    )
 set_property(TARGET airhost_shared PROPERTY POSITION_INDEPENDENT_CODE ON)
 
+add_library(libelf_pic STATIC IMPORTED)
```
Comment (jnider) [inline runtime_lib/airhost/CMakeLists.txt:45]: Also part of the libelf change

File: utils/github-clone-build-elfutils.sh
Comment (jnider) [inline utils/github-clone-build-elfutils.sh:1]: Also part of the libelf change

**Review by jnider — APPROVED** [review]
Other than the one question about libelf, everything else looks good

**Review by erwei-xilinx — APPROVED** [review]

File: .github/workflows/buildAndTest.yml (hunk 1 above)
Comment (eddierichter-amd) [reply .github/workflows/buildAndTest.yml:30]: I can edit the commit. Thanks for the comments I see how you are doing it in your mlir-aie fork.

Comment (eddierichter-amd) [reply .github/workflows/buildAndTest.yml]: In the moment felt better as a new commit so put in [04a27d4](https://github.com/Xilinx/mlir-air/pull/367/commits/04a27d44a94310b94a27d0ff068b683d38f215f9). Let me know what you think. I couldn't get CMake to automatically find LibElf if you know how to do that that could make it cleaner, but I am just pointing to the standard install location which also works.

Comment (jnider) [reply .github/workflows/buildAndTest.yml]: I don't know how to get CMake to search for a library. We could pass the path explicitly (like I was doing for the custom version) but with a default to the standard Ubuntu location. I think for now, relying on the standard Ubuntu location (as you are doing) is acceptable. If everyone will just use 'apt install' then it is highly unlikely the library will be at a different location.

Comment (eddierichter-amd) [reply .github/workflows/buildAndTest.yml]: 100% agree. Sweet will merge this in. Thanks!
